allennlp/commands/train.py

```diff
@@ -485,26 +445,136 @@ def _train_worker(
+
+    def finish(self, metrics: Dict[str, Any]):
+        if self.evaluation_dataset and self.evaluate_on_test:
             logger.info("The model will be evaluated using the best epoch weights.")
-            test_metrics = evaluate(
-                trainer.model,
-                evaluation_dataset,
-                evaluation_iterator,
-                cuda_device=trainer.cuda_device,
-                batch_weight_key=batch_weight_key,
+            test_metrics = training_util.evaluate(
+                self.model,
+                self.evaluation_dataset,
+                self.evaluation_iterator,
+                cuda_device=self.trainer.cuda_device,
+                batch_weight_key=self.batch_weight_key,
             )
 
             for key, value in test_metrics.items():
                 metrics["test_" + key] = value
-        elif evaluation_dataset:
+        elif self.evaluation_dataset:
             logger.info(
                 "To evaluate on the test set after training, pass the "
                 "'evaluate_on_test' flag, or use the 'allennlp evaluate' command."
             )
-        dump_metrics(os.path.join(serialization_dir, "metrics.json"), metrics, log=True)
+        common_util.dump_metrics(
+            os.path.join(self.serialization_dir, "metrics.json"), metrics, log=True
+        )
 
-    if not distributed:
-        return trainer.model
+    @classmethod
+    def from_partial_objects(
+        cls,
+        serialization_dir: str,
+        local_rank: int,
+        dataset_reader: DatasetReader,
+        train_data_path: str,
+        model: Lazy[Model],
+        iterator: DataIterator,
+        trainer: Lazy[TrainerBase],
+        vocabulary: Lazy[Vocabulary] = None,
+        datasets_for_vocab_creation: List[str] = None,
+        validation_dataset_reader: DatasetReader = None,
+        validation_data_path: str = None,
+        validation_iterator: DataIterator = None,
+        test_data_path: str = None,
+        evaluate_on_test: bool = False,
+        batch_weight_key: str = "",
+    ) -> "TrainModel":
+
+        datasets = training_util.read_all_datasets(
+            train_data_path=train_data_path,
+            dataset_reader=dataset_reader,
+            validation_dataset_reader=validation_dataset_reader,
+            validation_data_path=validation_data_path,
+            test_data_path=test_data_path,
+        )
 
-    return None  # to make mypy happy
+        if datasets_for_vocab_creation:
+            for key in datasets_for_vocab_creation:
+                if key not in datasets:
+                    raise ConfigurationError(f"invalid 'dataset_for_vocab_creation' {key}")
+
+        instance_generator = (
+            instance
+            for key, dataset in datasets.items()
+            if not datasets_for_vocab_creation or key in datasets_for_vocab_creation
+            for instance in dataset
+        )
+
+        vocabulary_ = vocabulary.construct(instances=instance_generator)
```

Comment:
I think having both the variable and the variable with the same name but with the underscore suffix is a bit confusing. What about renaming the former to `partial_vocabulary`?

Reply:
I agree that this isn't optimal. But we can't rename the method argument, because it has to match what we want to see in the config file.  We can only rename the constructed one.  Suggestions welcome.

Reply:
Oh, I see. What about renaming the other variable? To something like `constructed_vocabulary`?